File: Scripts/seth-1_scripts/gcc_preprocessing.R
library(shiny)
library(ggplot2)
library(reshape)
library(gplots)
library(RColorBrewer)
library(argparse)
library(ape)
library(RMySQL)
library(gridExtra)
library(Hmisc)

con <- dbConnect(MySQL(), user="setd", password="1234", dbname="aspminedb", host="192.38.13.196", port = 3306)

###########
# Summary #
###########




print("How many hitclusters are hit by a querycluster?")
# query <- "select family_hits from (select *, count(*) as family_hits from (select *,  pident*members/q_size as custom_score from (select *, count(*) as members from t_genecluster_bidir_hits where pident > 50 group by q_clust,clust_id) tx where members >= 0.5*q_size and members >= 0.25*h_size and q_size > 1) ty group by q_clust) tz;"
query <- "select family_hits, q_size, count(*) as fa_hit_count from (select *, count(*) as family_hits from (select *,  pident*members/q_size as custom_score from (select *, count(*) as members from t_genecluster_bidir_hits where pident > 50 group by q_clust,clust_id) tx where members >= 0.5*q_size and members >= 0.25*h_size and q_size >= 1) ty group by q_clust) tz group by family_hits, q_size;"
hits <- dbGetQuery(con, query)

#print(summary(hits))

hits <- subset(hits, family_hits>=2) # Decided to take the singletons out because they only match themselves

print(head(hits))

sub_hits <- subset(hits, family_hits>=10)

print(head(sub_hits))

###########################################################
# Abundance of multiple backbones in bidirectional hits ? #
###########################################################
print("How many gene clusters have multiple backbones?")

query <- "select buf3 as backbone, count(*) as amount from (select *, count(*) as members, sum(buf2) as buf3 from  (select *, case when q_sm_short != 'none' or h_sm_short != 'none' then  buf+1 else buf+0 end as buf2 from t_genecluster_bidir_hits where pident > 50) tw group by q_clust,clust_id) tx where members >= 0.25*q_size and members >= 0.25*h_size and q_size > 2 group by buf3;"

multi_b <- dbGetQuery(con, query)

print(multi_b)


# Shows the same just if backbone was found or not ;)

query <- "select buf3, count(*) from (select *, count(*) as members, max(buf2) as buf3 from  (select *, case when q_sm_short != 'none' or h_sm_short != 'none' then  buf+1 else buf+0 end as buf2 from t_genecluster_bidir_hits where pident > 50) tw group by q_clust,clust_id) tx where members >= 0.25*q_size and members >= 0.25*h_size and q_size > 2 group by buf3;"

has_b <- dbGetQuery(con, query)

print(has_b)



##############
# Singletons #
##############

query <- "select real_name, singleton  from (select org_id, count(*) as singleton from (select *, count(*) as family_hits from (select *,  pident*members/q_size as custom_score from (select *, count(*) as members from t_genecluster_bidir_hits where pident > 50 group by q_clust,clust_id) tx where members >= 0.5*q_size and members >= 0.25*h_size and q_size > 1) ty group by q_clust) tz where family_hits <= 1 and q_clust = clust_id group by org_id) ta join organism using(org_id);"

singletons <- dbGetQuery(con, query)

print(singletons)


# under construction query to fetch select q_clust from (select *, count(*) as family_hits from (select *,  pident*members/q_size as custom_score from (select *, count(*) as members from t_genecluster_bidir_hits where pident > 50 group by q_clust,clust_id) tx where members >= 0.5*q_size and members >= 0.25*h_size and q_size > 1) ty group by q_clust) tz where family_hits <= 1 and q_clust = clust_id;

# select distinct q_gene, org_id from t_genecluster_bidir_hits where q_clust = '9_8731_9';

# select * from (select distinct q_gene, go_term_id, q_sm_short from t_genecluster_bidir_hits as tgbh join protein_has_go as phg on q_gene = protein_id and tgbh.org_id = phg.org_id  where q_clust = '9_8731_9') tx join go using (go_term_id) order by q_gene;


############
# Printing #
############

print("Printing to Genecluster_family_preprocessing.pdf")

pdf('Genecluster_family_preprocessing.pdf')


#qplot(family_hits,data=hits, geom='histogram', binwidth=1, main = "Hitclusters matching to queryclusters")

#myPalette <- colorRampPalette(rev(brewer.pal(11, "Spectral")))
#sc <- scale_colour_gradient2(colours = myPalette(100), limits=c(1, 50))

#sc <- scale_fill_continuous(low = "#D0D1E6", high = "#023858",limits=c(1, 50), breaks = c(0,5,10,15,20,25,30,35,40,45,50))

qplot(family_hits, y=fa_hit_count, stat="identity", data=hits, geom="bar", fill=factor(q_size)) + guides(col = guide_legend(ncol = 3)) #theme(legend.text = element_text( size = 1))
#qplot(family_hits,data=sub_hits, geom='histogram', binwidth=1, main = "Hitclusters matching to queryclusters with more than 10 hits")
qplot(family_hits, y=fa_hit_count, stat="identity", data=sub_hits, geom="bar", fill=factor(q_size))

ggplot(singletons, aes(x = real_name, y = singleton))+
geom_bar(stat = "identity")+
coord_flip()

frame()

grid.table(multi_b)


dev.off()


#create table t_genecluster_testing as select *,  pident*members/q_size as custom_score from (select *, count(*) as members from t_genecluster_bidir_hits where pident > 50 group by q_clust,clust_id) tx where members >= 0.5*q_size and members >= 0.5*h_size and q_size > 2;

#create table t_gcc_testing as select * from (select *, count(*) as family_hits from (select *,  pident*members/q_size as custom_score from (select *, count(*) as members from t_genecluster_bidir_hits where pident > 50 group by q_clust,clust_id) tx where members >= 0.5*q_size and members >= 0.5*h_size and q_size > 1) ty group by q_clust) tz where family_hits >10;

File: Scripts/seth-1_scripts/gcc_preprocessing.py
################################
# Imports ######################
################################

import sys,os
sys.path.append("/home/seth/300asp/utils")
from aspmine_imports import *
import pandas as pd
from seth_imports import *



parser = argparse.ArgumentParser(description="Redo tables and stuff", usage="%(prog)s")
#parser.add_argument("--plot", "-p", required=False, default = False, help="Only execute plotting")
parser.add_argument("--redo_bidir", "-rb", required=False, action = 'store_true', help="Redo t_genecluster_bidir_hits table")
parser.add_argument("--redo_tables", "-rt", required=False, action = 'store_true', help="Redo tables ")
parser.add_argument("--singletons", "-s", required=False, action = 'store_true', help="Get singletons")

args = parser.parse_args()
redo_tables = args.redo_tables
redo_bidir = args.redo_bidir
singletons = args.singletons

cursor = asp_con(path='192.38.13.196', user='setd', pw='1234')


####################
# Table management #
####################

if redo_tables:
	query = "DROP TABLE IF EXISTS t_genecluster_testing_full"

	cursor.execute(query)

	query = "CREATE table t_genecluster_testing_full as\
	select *,  (buf3/2)+(pident*members/q_size/200) as custom_score from\
	(select *, count(*) as members, max(buf2) as buf3 from\
		(select *, case when q_sm_short != 'none' or h_sm_short != 'none' then  buf+1 else buf+0 end as buf2\
			from t_genecluster_bidir_hits) tw\ #deleted: where pident >50
			group by q_clust,clust_id) tx where members >= 0.25*q_size and members >= 0.25*h_size and q_size >= 2 and h_size >=2;"
	
# Can be used to count: select count(distinct(q_clust)), count(distinct(clust_id)) from t_genecluster_testing_full;

# select * from t_genecluster_testing where q_clust = clust_id and org_id != 3 and q_size != members; This finds out if some members are not found in a self hit

	cursor.execute(query)

	# old version: erase buf3/2 in the custom score and the 200 as well

	query = "DROP TABLE IF EXISTS t_gcc_testing_full"
	cursor.execute(query)
# Here a count does not really make sense since I group them by q_clust and h_clust gets picked randomly
	query = "CREATE table t_gcc_testing_full as\
	select * from (select *, count(*) as family_hits\
		from (select *,  pident*members/q_size as custom_score\
			from (select *, count(*) as members from t_genecluster_bidir_hits\
				group by q_clust,clust_id) tx\ # deleted: where pident > 50
				where members >= 0.25*q_size and members >= 0.25*h_size and q_size >= 2 and h_size >=2) ty group by q_clust) tz where family_hits >10;"
	cursor.execute(query)


# This is generating the initial bidirectional hits table between geneclusters
if redo_bidir:
	query = "DROP TABLE IF EXISTS t_genecluster_bidir_hits_full"

	cursor.execute(query)

	print "Deleted table t_genecluster_bidir_hits_full"

	query = "CREATE table t_genecluster_bidir_hits_full as\
	select tx.org_id, tx.clust_id as q_clust, tx.clust_size as q_size, tx.sm_protein_id as q_gene, tx.sm_short as q_sm_short, tx.h_seqkey as h_gene, ty.sm_short as h_sm_short, ty.clust_id, ty.clust_size as h_size, tx.bitscore, tx.pident\
	from t_antismash2blast as tx join t_antismash2blast as ty on tx.h_seqkey = ty.sm_protein_id where ty.h_seqkey = tx.sm_protein_id AND pident >= 50;"

	cursor.execute(query)

	print "Created new t_genecluster_bidir_hits_full table"



#########################
# Extracting Singletons #
#########################

# Singletons are only hitting themselves and nothing else, so they should be extracted before other tables are generated
if singletons:
	query = "select q_clust from (select *, count(*) as family_hits from (select *,  pident*members/q_size as custom_score from (select *, count(*) as members from t_genecluster_bidir_hits where pident > 50 group by q_clust,clust_id) tx where members >= 0.5*q_size and members >= 0.5*h_size and q_size > 1) ty group by q_clust) tz where family_hits <= 1 and q_clust = clust_id;"

	cursor.execute(query)
	singletons = list(cursor.fetchall())

	print singletons
